src/main/groovy/io/cify/framework/core/DeviceManager.groovy

```diff
@@ -135,13 +137,74 @@ class DeviceManager implements IDeviceManager {
             Device device = new Device(deviceId, category, desiredCapabilities)
             devices.add(device)
             setDeviceActive(device)
+        } catch (all) {
+            LOG.debug(MARKER, all.message, all)
+            throw new CifyFrameworkException("Failed to create device. Cause: $all.message")
+        }
+
+        addDeviceToTestReport(deviceId, category.toString())
+        capabilities.setAvailable(desiredCapabilities.getCapability(CAPABILITY_ID) as String, false)
+
+        return getActiveDevice(deviceId)
+    }
+
+    /**
+     * Creates device with capabilities identified by capabilityId
+     * @param capabilityId capability identifier
+     * @return Device
+     */
+    @Override
+    Device createDevice(String capabilityId) {
+        createDevice(capabilityId, generateRandomDeviceId())
+    }
+
+    /**
+     * Creates device with capabilities identified by capabilityId and with unique deviceId
+     * @param capabilityId capability identifier
+     * @param deviceId unique device id
+     * @return Device
+     */
+    @Override
+    Device createDevice(String capabilityId, String deviceId) {
```

Comment:
It has a lot in common with the same overloaded method above. Maybe it worth to move out the common part of these two methods? As for now I see code dups

Reply:
I've included it to "haveToBeRefactored" list